x-pack/plugins/security/server/authentication/index.test.ts

```diff
@@ -0,0 +1,350 @@
+    });
+
+    it('rejects with `Internal Server Error` when `authenticate` throws unhandled exception', async () => {
+      authenticate.mockRejectedValue(new Error('something went wrong'));
+
+      await authHandler(httpServerMock.createKibanaRequest(), mockAuthToolkit);
+
+      expect(mockAuthToolkit.rejected).toHaveBeenCalledTimes(1);
+      const [[error]] = mockAuthToolkit.rejected.mock.calls;
+      expect(error.message).toBe('something went wrong');
+      expect(getErrorStatusCode(error)).toBe(500);
+
+      expect(mockAuthToolkit.authenticated).not.toHaveBeenCalled();
+      expect(mockAuthToolkit.redirected).not.toHaveBeenCalled();
+    });
+
+    it('rejects with wrapped original error when `authenticate` fails to authenticate user', async () => {
+      const esError = Boom.badRequest('some message');
+      authenticate.mockResolvedValue(AuthenticationResult.failed(esError));
+
+      await authHandler(httpServerMock.createKibanaRequest(), mockAuthToolkit);
+
+      expect(mockAuthToolkit.rejected).toHaveBeenCalledTimes(1);
+      const [[error]] = mockAuthToolkit.rejected.mock.calls;
+      expect(error).toBe(esError);
+
+      expect(mockAuthToolkit.authenticated).not.toHaveBeenCalled();
+      expect(mockAuthToolkit.redirected).not.toHaveBeenCalled();
+    });
+
+    it('includes `WWW-Authenticate` header if `authenticate` fails to authenticate user and provides challenges', async () => {
+      const originalError = Boom.unauthorized('some message');
+      originalError.output.headers['WWW-Authenticate'] = [
+        'Basic realm="Access to prod", charset="UTF-8"',
+        'Basic',
+        'Negotiate',
+      ] as any;
+      authenticate.mockResolvedValue(AuthenticationResult.failed(originalError, ['Negotiate']));
+
+      await authHandler(httpServerMock.createKibanaRequest(), mockAuthToolkit);
+
+      expect(mockAuthToolkit.rejected).toHaveBeenCalledTimes(1);
+      const [[error]] = mockAuthToolkit.rejected.mock.calls;
+      expect(error.message).toBe(originalError.message);
+      expect((error as Boom).output.headers).toEqual({ 'WWW-Authenticate': ['Negotiate'] });
+
+      expect(mockAuthToolkit.authenticated).not.toHaveBeenCalled();
+      expect(mockAuthToolkit.redirected).not.toHaveBeenCalled();
+    });
+
+    it('returns `unauthorized` when authentication can not be handled', async () => {
+      authenticate.mockResolvedValue(AuthenticationResult.notHandled());
+
+      await authHandler(httpServerMock.createKibanaRequest(), mockAuthToolkit);
+
+      expect(mockAuthToolkit.rejected).toHaveBeenCalledTimes(1);
+      const [[error]] = mockAuthToolkit.rejected.mock.calls;
+      expect(error.message).toBe('Unauthorized');
+      expect(getErrorStatusCode(error)).toBe(401);
+
+      expect(mockAuthToolkit.authenticated).not.toHaveBeenCalled();
+      expect(mockAuthToolkit.redirected).not.toHaveBeenCalled();
+    });
+  });
+
+  describe('getServerBaseURL()', () => {
```

Comment:
__note:__ diff#5: added few tests to cover this new (well, old, since we don't need it in master anymore) function